src/events.rs

```diff
@@ -87,51 +86,24 @@ pub struct EventsDecoder<T> {
     marker: PhantomData<fn() -> T>,
 }
 
-impl<T: System> TryFrom<Metadata> for EventsDecoder<T> {
-    type Error = EventsError;
-
-    fn try_from(metadata: Metadata) -> Result<Self, Self::Error> {
-        let mut decoder = Self {
+impl<T: System> EventsDecoder<T> {
+    /// Creates a new `EventsDecoder`.
+    pub fn new(metadata: Metadata) -> Self {
+        Self {
             metadata,
             type_sizes: HashMap::new(),
             marker: PhantomData,
-        };
-        // register default event arg type sizes for dynamic decoding of events
-        decoder.register_type_size::<bool>("bool")?;
-        decoder.register_type_size::<u32>("ReferendumIndex")?;
-        decoder.register_type_size::<[u8; 16]>("Kind")?;
-        decoder.register_type_size::<[u8; 32]>("AuthorityId")?;
-        decoder.register_type_size::<u8>("u8")?;
-        decoder.register_type_size::<u32>("u32")?;
-        decoder.register_type_size::<u32>("AccountIndex")?;
-        decoder.register_type_size::<u32>("SessionIndex")?;
-        decoder.register_type_size::<u32>("PropIndex")?;
-        decoder.register_type_size::<u32>("ProposalIndex")?;
-        decoder.register_type_size::<u32>("AuthorityIndex")?;
-        decoder.register_type_size::<u64>("AuthorityWeight")?;
-        decoder.register_type_size::<u32>("MemberCount")?;
-        decoder.register_type_size::<T::AccountId>("AccountId")?;
-        decoder.register_type_size::<T::BlockNumber>("BlockNumber")?;
-        decoder.register_type_size::<T::Hash>("Hash")?;
-        decoder.register_type_size::<u8>("VoteThreshold")?;
```

Comment:
While this is an extremely imperfect solution, IIRC these type size registrations are required in case we receive events from modules not defined in `subxt` here: https://github.com/paritytech/substrate-subxt/blob/master/src/rpc.rs#L479 so we can skip over them. Of course with improvements to the metadata, potentially integrating `https://github.com/paritytech/scale-info` this would no longer be required. Also `desub` has an alternative approach which utilizes mappings from `polkadot-js` https://github.com/paritytech/desub. See https://github.com/paritytech/substrate-subxt/issues/72

Reply:
I see, I'll revert for now. scale-info sounds cool

Reply:
Done. Is there any eta on scale-info? Seems like that would solve a lot of problems

Reply:
I'm currently working on generics for a `v0.1` release, and once I've implemented associated type representation I will experiment with substrate integration.